File: gfg-weekly-coding-contest-109/Count Subarrays.cpp

class Solution {
  public:
    // Version 1: Hashmap
    // pre[val][count]: The number of val with count times of X until now
    // TC: O(n), SC: O(n)
    /*
    long long countSubarray(int N, vector<int> &arr, int X, int K) {
        // code here
        unordered_map<int, unordered_map<int, int>> pre;
        pre[X][0] = 1;
        long long M = pow(10, 9) + 7;
        long long ans = 0, current = 0;
        for (int i = 0; i < N; i++) {
            current += (arr[i] == X? 1: 0);
            ans = (ans + pre[arr[i]][current - K]) % M;
            pre[arr[i]][current]++;
        }
        return ans;
    }
    */
    
    // Verion 2: Sliding window
    // When we have k X, count the occurrence of non-X numbers from left.
    // Precede the right position until we meet X to count the number of valid substrings
    // TC: O(n), SC: O(n)
    long long countSubarray(int N, vector<int> &arr, int X, int K) {
        long long ans = 0;
        int count = 0;
        int l = 0, r = 0;
        int M = pow(10, 9) + 7;
        while (r < N) {
            if (arr[r] == X) count++;
            r++;
            if (count == K) {
                unordered_map<int, int> pre;
                while(arr[l] != X) {
                    pre[arr[l]]++;
                    l++;
                }
                
                ans = (ans + 1) % M;
                while(r < N && arr[r] != X) {
                    ans = (ans + pre[arr[r]]) % M;
                    r++;
                }
                l++;
                count--;
            }
        }
        return ans;
    }
};

File: gfg-weekly-coding-contest-109/Count Subarrays.java


class Solution{
    // Version 1: Hashmap
    // pre[val][count]: The number of val with count times of X until now
    // TC: O(n), SC: O(n)
    /*
    public int countSubarray(int arr[], int n, int X, int K){
        // Code here. 
        Map<Integer, Map<Integer, Integer>> pre = new HashMap<>();
        pre.put(X, new HashMap<Integer, Integer>());
        pre.get(X).put(0, 1);
        long ans = 0;
        int M = (int)Math.pow(10, 9) + 7;
        int current = 0;
        for (int i = 0; i < n; i++) {
            int val = arr[i];
            if (val == X) current++;
            Map<Integer, Integer> f;
            if (pre.containsKey(val)) {
                f = pre.get(val);
                if (current >= K && f.containsKey(current - K)) {
                    ans = (ans + f.get(current - K)) % M;
                }
            } else {
                f = new HashMap<Integer, Integer>();
            }
            f.put(current, f.getOrDefault(current, 0) + 1);
            pre.put(val, f);
        }
        return (int)ans;
    }
    */
    
    // Verion 2: Sliding window
    // When we have k X, count the occurrence of non-X numbers from left.
    // Precede the right position until we meet X to count the number of valid substrings
    // TC: O(n), SC: O(n)
    public int countSubarray(int arr[], int n, int X, int K){
        int ans = 0;
        int count = 0;
        int l = 0, r = 0;
        int M = (int)Math.pow(10, 9) + 7;
        Map<Integer, Integer> pre = new HashMap<>();
        while (r < n) {
            if (arr[r] == X) count++;
            r++;
            if (count == K) {
                pre.clear();
                while(arr[l] != X) {
                    pre.put(arr[l], pre.getOrDefault(arr[l], 0) + 1);
                    l++;
                }
                
                ans = (ans + 1) % M;
                while(r < n && arr[r] != X) {
                    if (pre.containsKey(arr[r])) {
                        ans = (ans + pre.get(arr[r])) % M;
                    }
                    r++;
                }
                l++;
                count--;
            }
        }
        return ans;
    }
    
}